enterprise/frontend/src/embedding-sdk/lib/interactive-question/run-question-on-navigate.ts

```diff
@@ -1,52 +1,57 @@
+import { runQuestionQuerySdk } from "embedding-sdk/lib/interactive-question/run-question-query";
 import type {
   NavigateToNewCardParams,
-  SdkQuestionResult,
+  SdkQuestionState,
 } from "embedding-sdk/types/question";
 import { loadCard } from "metabase/lib/card";
+import { loadMetadataForCard } from "metabase/questions/actions";
 import { getMetadata } from "metabase/selectors/metadata";
 import { getCardAfterVisualizationClick } from "metabase/visualizations/lib/utils";
 import Question from "metabase-lib/v1/Question";
 import { cardIsEquivalent } from "metabase-lib/v1/queries/utils/card";
 import type { Dispatch, GetState } from "metabase-types/store";
 
-import { runQuestionOnQueryChangeSdk } from "./run-question-on-query-change";
-
 interface RunQuestionOnNavigateParams extends NavigateToNewCardParams {
   originalQuestion?: Question;
+  onQuestionChange: (question: Question) => void;
 }
 
 export const runQuestionOnNavigateSdk =
   (params: RunQuestionOnNavigateParams) =>
   async (
     dispatch: Dispatch,
     getState: GetState,
-  ): Promise<SdkQuestionResult | null> => {
-    let { nextCard, previousCard, originalQuestion } = params;
+  ): Promise<SdkQuestionState | null> => {
+    let {
+      nextCard,
+      previousCard,
+      originalQuestion,
+      cancelDeferred,
+      onQuestionChange,
+    } = params;
 
     // Do not reload questions with breakouts when clicking on a legend item
     if (previousCard === nextCard) {
       return null;
     }
 
-    const metadata = getMetadata(getState());
-
     // Fallback when a visualization legend is clicked
     if (cardIsEquivalent(previousCard, nextCard)) {
       nextCard = await loadCard(nextCard.id, { dispatch, getState });
     } else {
       nextCard = getCardAfterVisualizationClick(nextCard, previousCard);
     }
 
-    const previousQuestion = new Question(previousCard, metadata);
-    const nextQuestion = new Question(nextCard, metadata);
+    // Optimistic update the UI before we re-fetch the query metadata.
+    onQuestionChange(new Question(nextCard, getMetadata(getState())));
+
+    await dispatch(loadMetadataForCard(nextCard));
```

Comment:
Updating the question before loading the metadata allows the UI to be snappy again.